### DISCUSSION TITLE
Performance issue with --tw-empty

### DISCUSSION THREAD

**Discussion by brunoabreu** [ben-rogerson/twin.macro#851] [General]
We are investigating a performance issue in Firefox with our app, which uses Next.js 13 (app router) + `emotion` with `twin.macro`.

We noticed that the default value for empty variables inserted by `twin.macro` (`var(--tw-empty,/*!*/ /*!*/)`) is causing a significant impact when the browser needs to repaint the layout. Using the default value from Tailwind, a blank space, the difference is dramatic:

| `var(--tw-empty,/*!*/ /*!*/)` | blank space |
| :-----: | :-----: |
|<b>459ms</b><img width="600" alt="309042122-fcd5d737-16a9-43be-b10a-576ad3e6ef3d" src="https://github.com/ben-rogerson/twin.macro/assets/2726345/7cdab21b-3128-4b6a-823d-070fc497ccad"> |<b>117ms</b><img width="429" alt="309042293-db161c78-d86a-4449-bd2b-17bba202b25c" src="https://github.com/ben-rogerson/twin.macro/assets/2726345/58922f35-f772-4a4a-b3d5-569cffc3f5cc"> |

Firefox seems to benefit the most, but a performance gain is also noticeable in Google Chrome.

| `var(--tw-empty,/*!*/ /*!*/)` | blank space |
| :-----: | :-----: |
|<b>88ms</b> <img width="408" alt="309044379-db761ab8-490f-4809-8eaa-2948f4815d87" src="https://github.com/ben-rogerson/twin.macro/assets/2726345/91cf9a2c-9f4d-466e-9692-eeb2d73c7500"> | <b>56ms</b> <img width="363" alt="309044467-c78b9f76-7808-4905-af02-56ae48668f9b" src="https://github.com/ben-rogerson/twin.macro/assets/2726345/f7a9a752-d5e5-453a-ae6b-de91f9bf039a">|

For now, we have created a [stylis](https://github.com/thysultan/stylis) plugin to replace `var(--tw-empty,/*!*/ /*!*/)` with a blank space, undoing what is done by `twin.macro`. Like this:

```ts
import { GlobalStyles as TwinGlobalStyles } from 'twin.macro'
import { CacheProvider } from '@emotion/react'
import createCache, { type StylisPlugin } from '@emotion/cache'

const TW_EMPTY = 'var(--tw-empty,/*!*/ /*!*/)'

function noTwEmptyPlugin(...[element]: Parameters<StylisPlugin>) {
  if (element.type === 'decl' && element.children === TW_EMPTY) {
    element.value = element.value.replace(TW_EMPTY, ' ')
  }
}

const cache = createCache({ key: 'notwempty', stylisPlugins: [noTwEmptyPlugin] })

export function GlobalStyles() {
  return <CacheProvider value={cache}><TwinGlobalStyles /></CacheProvider>
}
```

We also tried applying a patch using `yarn` to prevent this substitution that `twin.macro` makes for `--tw-empty`, but it is not so simple. Unfortunately, `emotion` (through `stylis`), removes the variables believing they are dead code.

The best solution probably involves modifying both `twin.macro` and `stylis`. But it would be helpful if `twin.macro` provided a configuration option to leave the blank space instead of making the substitution.

**Comment by ben-rogerson** [comment]
Hey there

I remember coming up with the `/*!*/ /*!*/` variable fallback workaround to get around the "dead code" shaving and didn't realise there was a performance hit when using it. 

If twin has a new option added to remove the space then the values still aren't read correctly and we end up with styles like this (from the [vite-emotion-typescript](https://github.com/ben-rogerson/twin.examples/tree/master/vite-emotion-typescript) example:

![image](https://github.com/ben-rogerson/twin.macro/assets/21288568/bccfbb1c-57e9-4ce3-a1fb-218e667f5294)

Do you know of a way to get Emotion to read those empty values? I haven't been able to find a suitable replacement.

**Reply by brunoabreu** [reply]
Hey Ben!

The idea is to stop using the `--tw-empty` var, and just leave the original whitespace. With our `stylis` plugin it looks like this:

<img width="383" alt="image" src="https://github.com/ben-rogerson/twin.macro/assets/2726345/f85ddcd9-bdf4-4e22-a448-78ad487f8cc1">

But sadly, if `twin.macro` outputs CSS like that, `stylis` will detect these properties as dead code and remove them, which also breaks Tailwind.

<img width="669" alt="image" src="https://github.com/ben-rogerson/twin.macro/assets/2726345/e24396a3-5c42-42c2-8f49-cdb974a714a3">

I've opened an [issue](https://github.com/thysultan/stylis/issues/328) in the `stylis` repository. Once it gets fixed, we can add an option to make `twin.macro` stop replacing the whitespace values for these custom properties.
